ARCH/X86: Introduce non temporal buffer transfer

arun-chandran-edarath opened the pull request:
Processors based on AMD's Zen 3/4 architecture typically organize CPU cores into clusters of 8 or more cores. Within each cluster, these cores share a unified L3 cache, optimizing data transfer speeds among them. However, when data needs to travel between cores located in different clusters that do not share a common L3 cache, the transfer speed can be significantly reduced, potentially becoming 2-3 orders of magnitude slower.

To leverage this microarchitecture effectively
MPI applications often employ 'n-ranks X m-threads' configuration.

   n <= Number of unified L3caches in system
   m >= Number of cores within a cluster

When UCX is used as the underlying data transfer mechanism, depending on the rendezvous threshold two ranks/processes use either CICO(copy-in copy-out) or SCOPY(single copy) to move data between them in an intra-node scenario.
This patch improves the data transfer speed/application-performance on AMD's Zen 3/4 architecture for both CICO and SCOPY when sender and receiver processes are pinned to cores that don't share a common L3 cache.

a)CICO

              Sender                      Receiver
         +-------------+               +------------+
         +.............+               +............+
         +    sbuf     +               +   rbuf     +
         +.............+               +............+
         +             +               +            +
         +             +               +            +
         +             +               +            +
         +.............+>>>>>>>>>>>>>>>+............+
         +   rshmem    +  POSIX/SYSV   +            +
         +   mapped    +   mapping     +  rshmem    +
         +   area      +               +            +
         +.............+<<<<<<<<<<<<<<<+............+
         +             +               +            +
         +-------------+               +------------+

In the current method, the sender process first moves the data from source buffer(sbuf) to the remote shared memory (rshmem) using the glibc's memcpy and informs the receiver process about the presence of new data. The receiver process then proceeds to copy the data from rshmem to its local buffer(rbuf) once again utilizing glibc's memcpy, thus completing the transfer.

In contrast, the new approach optimizes this by using distinct memory copy techniques. It utilizes
'nontemporal store' or 'store with PREFETCHNTA' when copying data from sbuf to rshmem and employs 'loads with PREFETCHNTA' when copying data from rshmem to rbuf. This optimization effectively reduces data transfer latency by circumventing cache-to-cache data transfers, which tend to be slower, especially when cores are situated in different clusters. Additionally, this method helps in minimizing cache pollution on both the sender and receiver sides,
further enhancing overall performance.

b)SCOPY (only xpmem is considered as other methods use memcpy inside kernel)

              Sender                          Receiver
         +-------------+               +--------------------+
         +             +               +....................+
         +             +               +       rbuf         +
         +             +               +....................+
         +             +               +                    +
         +             +               +                    +
         +             +               +                    +
         +.............+>>>>>>>>>>>>>>>+....................+
         +    sbuf     +               +   sbuf_xpmem_map   +
         +.............+>>>>>>>>>>>>>>>+                    +
         +             +               + physical pages of  +
         +             +               + sbuf mapped to     +
         +             +               + receiver's virtual +
         +             +               + memory by fault    +
         +             +               + handler in xpmem.ko+
         +             +               +....................+
         +             +               +                    +
         +             +               +                    +
         +-------------+               +--------------------+

The new method replaces glibc's memcpy of the receiver's data from sbuf_xpmem_map (src) to rbuf (dst) with 'loads with PREFETCHNTA'. This minimizes cache pollution inside receiver's core(sbuf pages are mapped with rw flags in the receiver process).


Acked-by: Edgar Gabriel <edgar.gabriel@amd.com>


How to build
---------------
use "--enable-builtin-memcpy=no  --enable-optimizations --enable-nt-buffer-transfer" while configuring  

Eg:
$./contrib/configure-release --prefix=/install_path --with-xpmem=/xpmem_path  --enable-builtin-memcpy=no  --enable-optimizations --enable-nt-buffer-transfer

How to run
-------------
During mpirun nt-buffer-transfer is dynamically selected according to the table below.

![ntbt_selection_criteria](https://github.com/openucx/ucx/assets/146521248/48b30c71-92e4-4296-9f74-571296c2abe2)


use "-x UCX_NT_BUFFER_TRANSFER_MIN=0" during mpirun for a workload that doesn't have more than one rank in the L3 domain

Below is the comparison of latency when two ranks are in 2 different numa domains (From a Milan:Zen3 node)
-------------------------------------------------------------------------------------------------------------------------

cmd: mpirun -np 2 --map-by numa --bind-to core --mca pml ucx  --mca opal_common_ucx_tls any --mca opal_common_ucx_devices any -x UCX_TLS=^ib -x UCX_NT_BUFFER_TRANSFER_MIN=0 ./osu_latency -m 1:268435456 -i 50000 -x 5000

![latancy_compare](https://github.com/openucx/ucx/assets/146521248/42f04dee-ef1c-4c93-bf4a-cf8e9a574d87)


**_Note: The default threshold values for selecting the nt-buffer-transfer for an individual transfer are set at 3/4 the size of L3 cache to address MPI Binding, where multiple ranks share a common L3 cache. If the MPI workload has only one rank in L3, nt-buffer-transfer can be used for all lengths, and the user can enforce this by passing "-x UCX_NT_BUFFER_TRANSFER_MIN=0" in the command line. ._**

--Arun

yosefe commented on src/ucp/rndv/rndv_rkey_ptr.c:
```diff
@@ -135,8 +136,15 @@ static unsigned ucp_proto_rndv_progress_rkey_ptr(void *arg)
     ucp_trace_req(req, "rkey_ptr unpack %zd from %p at offset %zd/%zd",
                   seg_size, src, offset, length);
 
+#ifdef ENABLE_AMD_BUFFER_TRANSFER
```
can we avoid the compile-time option, and select the memcpy implementation according to the runtime CPU detection?
see https://github.com/openucx/ucx/blob/2d6595440b124df0631d7ef60f6bbde036ef55db/src/ucs/arch/x86_64/global_opts.h#L23-L23 as example

arun-chandran-amd replied:
@yosefe Thanks for the review. In __amd_buffer_transfer_method__ I need to identify the direction of memory transfer, I explained it in the attached figure (shared_memory_transfer.jpg) for the copy-in-copy-out case.  Is this way of identifying the transfer direction acceptable? or Can I get it as an input in ucs_memcpy_relaxed(void *dst, const void *src, size_t len, int direction)?
So in short this is not just another version of memcpy but rather a direction-aware variant. Will other architectures be interested in this method?


![shared_memory_transfer](https://github.com/openucx/ucx/assets/146521248/daa3777a-8316-48c8-9e0e-de2c8768dec7)

yosefe replied:
Hi @arun-chandran-amd , we can definitely extend ucs_memcpy_relaxed to get more parameters that could improve performance, even if some architectures do not benefit from it.

arun-chandran-amd replied:
@yosefe Do you have any ideas/pointers on how to implement this? Can multiple calls to ucp_tag_send_xxx from the threads inside a process/rank will result in simultaneous calls to  ucs_memcpy_relaxed() in the sender/receiver?

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +707,257 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+#define YMM_SZ      32
+
+static UCS_F_ALWAYS_INLINE
+void* amd_copy_to_mapped_area(void *dst, const void *src, size_t len)
```
pls use ucs_cpu_ prefix

arun-chandran-amd replied:
Will change the function names

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +707,257 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+
+    /* First handle lengths that fall usually within eager short range */
+    if (ucs_likely(len <= 128)) {
+        switch (_lzcnt_u32(len))
+        {
+            case 32: /* 0 */
```
code style

arun-chandran-amd replied:
Will change the style

yosefe commented on src/ucs/configure.m4:
```diff
@@ -300,6 +300,23 @@ AS_IF([test "x$enable_builtin_memcpy" != xno],
+#
+# Enable amd optimized buffer transfer
+#
+AC_ARG_ENABLE([amd-buffer-transfer],
+	AS_HELP_STRING([--enable-amd-buffer-transfer],
+	               [Enable amd optimized buffer transfer routine, default: NO]),
+	[],
+	[enable_amd_buffer_transfer=no])
+
+AS_IF([test "x$enable_amd_buffer_transfer" = xyes],
+	  [AS_MESSAGE([enabling amd optimized buffer transfer])
+	   AC_DEFINE([ENABLE_AMD_BUFFER_TRANSFER], [1], [Enable amd optimizedi buffer transfer])
+	   HAVE_AMD_BUFFER_TRANSFER=yes],
+	  [:]
+  )
```
why not always enable it in compile-time?

arun-chandran-amd replied:
This method will not be suitable for transfers between cores that share a common L3cache, that is the reason I kept it as 'disabled' by default. 

Or can  I do like this?

a) Enable this feature always at compile time; Always disable at runtime
b) Use a ucx command-line variable (EG: -x UCX_ENABLE_DIRECTION_AWARE_COPY) to enable runtime.

arun-chandran-edarath commented:
@yosefe I have re-designed it by adding the copy direction in ucx_memcpy_relaxed(). Please see.

Dynamic controls with variables similar to builtin_memcpy_min & builtin_memcpy_max have not yet implemented; I will add them soon.
--Arun

yosefe reviewed (commented):
pls fix the commit title , see CI failure (would need force push)

arun-chandran-edarath commented:
> pls fix the commit title , see CI failure (would need force push)

Done. I have forcefully pushed after changing the commit titles of all the commits in this series.

arun-chandran-edarath commented on src/ucs/arch/generic/cpu.h:
```diff
@@ -11,6 +11,11 @@
+typedef enum {
+    BUFF_NORMAL = 0,
+    BUFF_NT_SEND,
+    BUFF_NT_RECV
```
Done. Thanks for the nice suggestion, looks better now

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +707,261 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+ * application can choose the cache hotness of the final buffer
+ */
+void ucs_x86_nt_buffer_transfer(void *dst, const void *src, size_t len,
+                                buff_transfer_t type)
```
Changed the enum name.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +707,261 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+
+    if (ucs_likely(len <= 128)) {
+        goto copy_bytes_le_128;
+    } else {
```
Done.

arun-chandran-edarath commented on src/ucs/sys/iovec.c:
```diff
@@ -40,10 +42,11 @@ size_t ucs_iov_copy(const struct iovec *iov, size_t iov_cnt,
+
         if (dir == UCS_IOV_COPY_FROM_BUF) {
-            memcpy(iov_buf, UCS_PTR_BYTE_OFFSET(buf, copied), len);
+            ucs_memcpy_relaxed(iov_buf, UCS_PTR_BYTE_OFFSET(buf, copied), len, BUFF_NT_RECV);
```
Modified.

arun-chandran-edarath commented on src/uct/sm/mm/base/mm_ep.c:
```diff
@@ -320,7 +342,7 @@ static UCS_F_ALWAYS_INLINE ssize_t uct_mm_ep_am_common_send(
     case UCT_MM_SEND_AM_SHORT:
         /* write to the remote FIFO */
-        uct_am_short_fill_data(elem + 1, header, payload, length);
+        uct_mm_ep_short_fill_data(elem + 1, header, payload, length);
```
Initially, I thought not to modify uct_am_short_fill_data() because other UCTs used it.
Adding a hint param should be fine, because with hint as '0' it will fallback to memcpy().

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +707,261 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
     return cache_count == UCS_CPU_CACHE_LAST ? UCS_OK : UCS_ERR_UNSUPPORTED;
 }
 
+#ifdef ENABLE_NT_BUFFER_TRANSFER
```
Non-temporal-buffer transfer is best suited to MPI workloads with a single rank within a CCD.
Eg: 16 ranks on a Milan node (Total 128 cores organized into 8 CCDs). For this type of workload all the memory transfers are between cores/processes located in different L3caches

This method will add more delays to transfers between cores that share a common L3cache, that is the reason I kept it as 'disabled' by default,  to make sure that those who enable it know what they are doing.

Alternatively a dynamic runtime enablement similar to the usage of builtin_memcpy_min & builtin_memcpy_max can be used to make it on/off at runtime.

arun-chandran-edarath commented:
@yosefe I am not sure why the builds are failing in the automated tests, on my system I don't get such failures if I build without enabling 'nt-buffer-transfer'. Could you please help to solve it?

yosefe commented:
> @yosefe I am not sure why the builds are failing in the automated tests, on my system I don't get such failures if I build without enabling 'nt-buffer-transfer'. Could you please help to solve it?

This is the error:
```
checking that generated files are newer than configure... done
configure: error: conditional "HAVE_NT_BUFFER_TRANSFER" was never defined.
Usually this means the macro was only invoked conditionally.
```

Maybe remove HAVE_NT_BUFFER_TRANSFER altogether?

arun-chandran-edarath commented:
> > @yosefe I am not sure why the builds are failing in the automated tests, on my system I don't get such failures if I build without enabling 'nt-buffer-transfer'. Could you please help to solve it?
> 
> This is the error:
> 
> ```
> checking that generated files are newer than configure... done
> configure: error: conditional "HAVE_NT_BUFFER_TRANSFER" was never defined.
> Usually this means the macro was only invoked conditionally.
> ```
> 
> Maybe remove HAVE_NT_BUFFER_TRANSFER altogether?

It was missing in configure.ac. I fixed it and pushed, let us see how the build behaves now. Thanks.

yosefe commented:
@arun-chandran-edarath pls see the merge conflicts

arun-chandran-edarath commented:
@yosefe I have implemented a dynamic-range-based usage depending on the total length of the transfer. Please take a look.

arun-chandran-edarath commented:
How to handle the build requirements?

Building it requires the use of --enable-optimizations (which in turn selects the necessary -march and -mavx options). How to specify if 'nt-buffer-transfer' is enabled and then select the extra compiler flags?

arun-chandran-edarath commented:
Hi All,

I want to collaborate with the folks from intel to get their opinion on setting the default values for all the tunable parameters of the nt-buffer-transfer patch. @yosefe Could you please help me with this request?

--Arun

arun-chandran-edarath commented:
@yosefe 
This PR has been sitting idle for a while, can you consider merging it before we branch out for 1.17?
  
Please let me know if there is anything I can do to help move this forward, or if there are any concerns or issues that need to be addressed.

--Arun

yosefe reviewed (commented):
in general, to expedite the review process, I'd suggest to break down this PR into 2 parts:
1. adding hint and total_len to ucs_memcpy_relaxed and adjusting upper layers accordingly
2. adding read/write prefetch and implementing NT memory copy in UCS

yosefe commented:
/azp run perf

arun-chandran-edarath commented:
> 2\. implementing



> in general, to expedite the review process, I'd suggest to break down this PR into 2 parts:
> 
> 1. adding hint and total_len to ucs_memcpy_relaxed and adjusting upper layers accordingly
> 2. adding read/write prefetch and implementing NT memory copy in UCS

@yosefe, sorry for the delay in my response. I was on vacation.

As per your suggestion, I have broken down the PR into three parts:

1) Adding total_len and hint to ucs_memcpy_relaxed.
2) Adding the prefetch functions required for nt-buffer-transfer().
3) Implementing nt-buffer-transfer() for x86_64.
To expedite the review process, I have implemented parts 1 and 2 and pushed the changes. I have addressed most of the comments related to parts 1 and 2 in this push. However, there are some comments that require a more in-depth knowledge of the UCX codebase, which I currently lack. I might need expert help to solve those.

I will work on those comments for part 3 and push the changes as soon as possible.

In the meantime, let's focus on closing parts 1 and 2.
--Arun

arun-chandran-edarath commented on src/tools/info/sys_info.c:
```diff
@@ -25,6 +25,7 @@
 static double measure_memcpy_bandwidth(size_t size)
 {
     ucs_time_t start_time, end_time;
+    ucs_arch_memcpy_hint_t hint = UCS_ARCH_MEMCPY_NT_NONE;
```
Removed

arun-chandran-edarath commented on src/ucs/arch/generic/cpu.h:
```diff
@@ -11,6 +11,11 @@
+typedef enum {
+    UCS_ARCH_MEMCPY_NT_NONE = 0,
+    UCS_ARCH_MEMCPY_NT_SOURCE = UCS_BIT(0),
+    UCS_ARCH_MEMCPY_NT_DEST = UCS_BIT(1),
```
Done.

yosefe replied:
hi, i think better rename the hint names according to the comment above

arun-chandran-edarath commented on src/uct/sm/mm/base/mm_iface.c:
```diff
@@ -2,6 +2,7 @@
  * Copyright (c) UT-Battelle, LLC. 2014-2015. ALL RIGHTS RESERVED.
  * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2001-2021. ALL RIGHTS RESERVED.
  * Copyright (C) Huawei Technologies Co., Ltd. 2023. ALL RIGHTS RESERVED.
+ * Copyright (C) Advanced Micro Devices, Inc. 2023. ALL RIGHTS RESERVED.
```
Removed.

arun-chandran-edarath commented on src/uct/sm/self/self.c:
```diff
@@ -302,6 +302,7 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_self_ep_t, uct_ep_t);
 ucs_status_t uct_self_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
                                   const void *payload, unsigned length)
 {
+    ucs_arch_memcpy_hint_t hint = UCS_ARCH_MEMCPY_NT_NONE;
```
Removed

yosefe commented on src/uct/ugni/smsg/ugni_smsg_ep.c:
```diff
@@ -233,7 +234,7 @@ uct_ugni_smsg_ep_am_common_send(uct_ugni_smsg_ep_t *ep, uct_ugni_smsg_iface_t *i
                                        const void *payload, unsigned length)
 {
-
+    ucs_arch_memcpy_hint_t hint = UCS_ARCH_MEMCPY_NT_NONE;
```
remove this var

arun-chandran-edarath commented on src/ucp/dt/dt_iov.c:
```diff
@@ -39,7 +39,7 @@ void ucp_dt_iov_gather(ucp_worker_h worker, void *dest, const ucp_dt_iov_t *iov,
                            UCS_PTR_BYTE_OFFSET(iov[*iovcnt_offset].buffer,
                                                *iov_offset),
-                           item_len_to_copy, mem_type);
+                           item_len_to_copy, mem_type, item_len);
```
@yosefe 
Could you please elaborate on how to extract total_len for ucp_dt_iov_gather() and ucp_dt_iov_scatter()?
--Arun

yosefe replied:
it would be additional parameter to these functions, passed from datatype iter - dt_iter->length

arun-chandran-edarath replied:
Like this?

`diff --git a/src/ucp/dt/datatype_iter.inl b/src/ucp/dt/datatype_iter.inl
index 0314d62a7..47fa02cc4 100644
--- a/src/ucp/dt/datatype_iter.inl
+++ b/src/ucp/dt/datatype_iter.inl
@@ -391,7 +391,8 @@ ucp_datatype_iter_next_pack(const ucp_datatype_iter_t *dt_iter,
                               dt_iter->type.iov.iov, length,
                               &next_iter->type.iov.iov_offset,
                               &next_iter->type.iov.iov_index,
-                              (ucs_memory_type_t)dt_iter->mem_info.type);
+                              (ucs_memory_type_t)dt_iter->mem_info.type,
+                              dt_iter->length);
         break;
     case UCP_DATATYPE_GENERIC:
         if (max_length != 0) {
diff --git a/src/ucp/dt/dt.c b/src/ucp/dt/dt.c
index f340fa42b..b6bf63297 100644
--- a/src/ucp/dt/dt.c
+++ b/src/ucp/dt/dt.c
@@ -120,7 +120,7 @@ size_t ucp_dt_pack(ucp_worker_h worker, ucp_datatype_t datatype,
     case UCP_DATATYPE_IOV:
         UCS_PROFILE_CALL_VOID(ucp_dt_iov_gather, worker, dest, src, length,
                               &state->dt.iov.iov_offset,
-                              &state->dt.iov.iovcnt_offset, mem_type);
+                              &state->dt.iov.iovcnt_offset, mem_type, length);
         result_len = length;
         break;
 
diff --git a/src/ucp/dt/dt_iov.c b/src/ucp/dt/dt_iov.c
index ea51b1d86..4bc365c2e 100644
--- a/src/ucp/dt/dt_iov.c
+++ b/src/ucp/dt/dt_iov.c
@@ -25,7 +25,7 @@
 
 void ucp_dt_iov_gather(ucp_worker_h worker, void *dest, const ucp_dt_iov_t *iov,
                        size_t length, size_t *iov_offset, size_t *iovcnt_offset,
-                       ucs_memory_type_t mem_type)
+                       ucs_memory_type_t mem_type, size_t total_len)
 {
     size_t length_it = 0;
     size_t item_len, item_reminder, item_len_to_copy;
@@ -40,7 +40,7 @@ void ucp_dt_iov_gather(ucp_worker_h worker, void *dest, const ucp_dt_iov_t *iov,
         ucp_dt_contig_pack(worker, UCS_PTR_BYTE_OFFSET(dest, length_it),
                            UCS_PTR_BYTE_OFFSET(iov[*iovcnt_offset].buffer,
                                                *iov_offset),
-                           item_len_to_copy, mem_type, item_len);
+                           item_len_to_copy, mem_type, total_len);
         length_it += item_len_to_copy;
 
         ucs_assert(length_it <= length);
diff --git a/src/ucp/dt/dt_iov.h b/src/ucp/dt/dt_iov.h
index 6f990a70b..fc7cf63dd 100644
--- a/src/ucp/dt/dt_iov.h
+++ b/src/ucp/dt/dt_iov.h
@@ -55,7 +55,7 @@ static inline size_t ucp_dt_iov_length(const ucp_dt_iov_t *iov, size_t iovcnt)
  */
 void ucp_dt_iov_gather(ucp_worker_h worker, void *dest, const ucp_dt_iov_t *iov,
                        size_t length, size_t *iov_offset, size_t *iovcnt_offset,
-                       ucs_memory_type_t mem_type);
+                       ucs_memory_type_t mem_type, size_t total_len);
 

`

arun-chandran-edarath replied:
I will push a temp commit please check whether I got it right, I thought it required addition of lengths from individual 'iovs'

yosefe replied:
yes, it looks good

arun-chandran-edarath replied:
Will push a new commit with this change.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -588,6 +588,30 @@ static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
 }
 #endif
 
+#if ENABLE_NT_BUFFER_TRANSFER
```
I've implemented it as a conditional to prevent the unnecessary use of two additional 64-bit variables when "NT_BUFFER_TRANSFER" is disabled. Please refer to the following definitions and structure declarations for more details
###########
#if ENABLE_NT_BUFFER_TRANSFER
#define UCS_ARCH_GLOBAL_OPTS_INITALIZER { \
    .builtin_memcpy_min     = UCS_MEMUNITS_AUTO, \
    .builtin_memcpy_max     = UCS_MEMUNITS_AUTO, \
    .nt_buffer_transfer_min = UCS_MEMUNITS_AUTO, \
    .nt_dest_threshold      = UCS_MEMUNITS_AUTO  \
}
#else
#define UCS_ARCH_GLOBAL_OPTS_INITALIZER { \
    .builtin_memcpy_min     = UCS_MEMUNITS_AUTO, \
    .builtin_memcpy_max     = UCS_MEMUNITS_AUTO  \
}
#endif

/* built-in memcpy & nt-buffer-transfer config */
typedef struct ucs_arch_global_opts {
    size_t builtin_memcpy_min;
    size_t builtin_memcpy_max;
#if ENABLE_NT_BUFFER_TRANSFER
    size_t nt_buffer_transfer_min;
    size_t nt_dest_threshold;
#endif
} ucs_arch_global_opts_t;
###########
If you believe this is excessive, I can modify it so that these are always compiled. What are your thoughts on this?

The ERMS memcpy instructions behave differently on various architectures and also depend on the length of the transfer. We should have named it ERMS_MEMCPY instead of BUILTIN_MEMCPY :). IMO these two methods should coexist, and the user should have the flexibility to use both simultaneously depending on the transfer length.(This would require platform-specific tuning to determine which method performs best for a given range)

yosefe replied:
i think adding these variables is not an issue, and better make the code simpler and reduce the scope of preprocessor ifdef to the minimum

arun-chandran-edarath replied:
Ok. Thanks for the clarification, I will remove this.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -588,6 +588,30 @@ static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
+        return user_val;
+    }
+
+    if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 2) * 3);
```
Given the limited time frame before the 1.17 branching, I prefer postponing this modification. I will address it later.

tvegas1 replied:
detail: multiply before divide?

tvegas1 replied:
eventually, why using a 3/4 factor, and not using all of the cache (4/4 factor) as a criterion? i guess if we trash whole cache with size > l3_cache, the perf drawback when ranks are on the same L3 would not even be there?

arun-chandran-edarath replied:
> detail: multiply before divide?

I wanted approximate values, so the shift operators are ok to be used in these calculations. 
You are right multiplying before the divide will give a more accurate approximation, I will change it.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,324 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
     return cache_count == UCS_CPU_CACHE_LAST ? UCS_OK : UCS_ERR_UNSUPPORTED;
 }
 
+#if (ENABLE_NT_BUFFER_TRANSFER) && (__AVX__)
```
Done. Thanks

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,324 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+
+static UCS_F_ALWAYS_INLINE
+size_t ucs_cpu_nt_dst_buffer_transfer(void *dst, const void *src, size_t len,
+                                      int hint, size_t total_len)
```
I kept it as 'int' because I am adding (oring) UCS_ARCH_MEMCPY_NT_DEST to the hint based on total_len.

####
void ucs_x86_nt_buffer_transfer(void *dst, const void *src, size_t len,
                                ucs_arch_memcpy_hint_t hint, size_t total_len)
{
    size_t tail_bytes;

    if (ucs_likely(len <= 128)) {
        goto copy_bytes_le_128;
    }

    /* Give precedence to non temporal destination */
    if (ucs_unlikely(total_len > ucs_global_opts.arch.nt_dest_threshold)) {
        hint |= UCS_ARCH_MEMCPY_NT_DEST;
    }

    if (hint & UCS_ARCH_MEMCPY_NT_DEST) {
        tail_bytes = ucs_x86_nt_dst_buffer_transfer(dst, src, len, hint, total_len);
###

yosefe replied:
so let's make it unsiged because it's flags

arun-chandran-edarath replied:
Done.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,324 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+#define YMM_SZ      32
+
+#ifndef __m256i_u
+typedef long long __m256i_u __attribute__((__vector_size__(32), __aligned__(1)));
```
Removed. It was added because of failures observed with older version (21.2) of 'nvc' during presubmit tests.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,324 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+            }
+
+            offset += 4 * 32;
+            len -= 128;
```
Done.

yosefe commented:
@arun-chandran-edarath pls avoid force push during review, fix the comments only by adding new commits or merge commits

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,324 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+#endif
+
+static UCS_F_ALWAYS_INLINE
+size_t ucs_cpu_nt_dst_buffer_transfer(void *dst, const void *src, size_t len,
```
Done.

arun-chandran-edarath commented:
> @arun-chandran-edarath pls avoid force push during review, fix the comments only by adding new commits or merge commits

Thanks for the heads up. I'll avoid force push during review from now on.

yosefe commented on src/ucp/core/ucp_am.c:
```diff
@@ -1350,7 +1350,7 @@ ucp_am_copy_data_fragment(ucp_recv_desc_t *first_rdesc, void *data,
     UCS_PROFILE_NAMED_CALL("am_memcpy_recv", ucs_memcpy_relaxed,
                            UCS_PTR_BYTE_OFFSET(first_rdesc + 1, offset),
-                           data, length);
+                           data, length, UCS_ARCH_MEMCPY_NT_SOURCE, length);
```
in this case the data would most likely be present in L3. but for TCP transport, for example, or when shared memory peer is on the same NUMA / L3 domain, it also can be true.

arun-chandran-edarath replied:
@yosefe how do we proceed on this?

yosefe replied:
from my perspective we don't have to do it in this PR

arun-chandran-edarath replied:
@yosef If there are no pending items in this PR, could you please consider merging it?

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,319 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
     return cache_count == UCS_CPU_CACHE_LAST ? UCS_OK : UCS_ERR_UNSUPPORTED;
 }
 
+#if ((ENABLE_NT_BUFFER_TRANSFER) && (__AVX__))
```
so i've meant here that we can remove all ( )..

arun-chandran-edarath replied:
There is one more reason I did it, I just mimicked the #if around "ENABLE_BUILTIN_MEMCPY" :)

`#if ENABLE_BUILTIN_MEMCPY
static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
{
    if (user_val != UCS_MEMUNITS_AUTO) {
        return user_val;
    }

    if (((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_INTEL) &&
         (ucs_arch_get_cpu_model() >= UCS_CPU_MODEL_INTEL_HASWELL)) ||
        (ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD) ||
        (ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_ZHAOXIN)) {
        return auto_val;
    } else {
        return UCS_MEMUNITS_INF;
    }
}
#endif
`

arun-chandran-edarath replied:
Done.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,319 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
 
+#if ((ENABLE_NT_BUFFER_TRANSFER) && (__AVX__))
+
+#define SWITCH_TO_NTSTORE_SZ (1464)
```
can we make it "const size_t ..." inside ucs_x86_nt_dst_buffer_transfer?

arun-chandran-edarath replied:
Done.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,319 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+    ucs_nt_write_prefetch(dst);
+    /* copy enough to make destination address 32 byte aligned */
+    offset = 64 - ((uintptr_t)dst & 0x1f);
+    y2 = _mm256_loadu_si256(src);
```
pls align = on column

arun-chandran-edarath replied:
Done.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,319 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+                _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 96), y3);
+
+                if (ucs_unlikely(hint & UCS_ARCH_MEMCPY_NT_SOURCE)) {
+                    ucs_nt_read_prefetch(UCS_PTR_BYTE_OFFSET(src, offset + 3 * 64));
```
pls use ( ) around * expressions - multiple locations

arun-chandran-edarath replied:
Done.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,319 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+        }
+
+        /* make the writes visible to the other core */
+        _mm_sfence();
```
is there a UCS macro for this?

arun-chandran-edarath replied:
Yea there is one "ucs_memory_bus_store_fence()" I missed it somehow :)

arun-chandran-edarath replied:
Done.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +736,324 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
+    } else if (hint & UCS_ARCH_MEMCPY_NT_SOURCE) {
+        tail_bytes = ucs_cpu_nt_src_buffer_transfer(dst, src, len);
+    } else {
+        memcpy(dst, src, len);
+        tail_bytes = 0;
```
WDYT?

arun-chandran-edarath replied:
I did it to keep the return path common to give the compiler more room for more optimization

yosefe commented on src/ucs/arch/x86_64/cpu.h:
```diff
@@ -112,6 +115,14 @@ static inline void *ucs_memcpy_relaxed(void *dst, const void *src, size_t len,
     }
 #endif
+
+#if ((ENABLE_NT_BUFFER_TRANSFER) && (__AVX__))
```
can remove all ( ) here

arun-chandran-edarath replied:
Removing the entire "#if ((ENABLE_NT_BUFFER_TRANSFER) && (__AVX__))" will make the normal compilation to fail. For nt-buffer-transfer to get compiled we need the '-mavx' flag.

I need at least "#if __AVX__" .

yosefe replied:
check only AVX condition

yosefe commented on src/ucs/configure.m4:
```diff
@@ -300,6 +300,21 @@ AS_IF([test "x$enable_builtin_memcpy" != xno],
+#
+# Enable non temporal buffer transfer
+#
+AC_ARG_ENABLE([nt-buffer-transfer],
+	AS_HELP_STRING([--enable-nt-buffer-transfer],
+	               [Enable non temporal buffer transfer routine, default: YES]),
+	[],
+	[enable_nt_buffer_transfer=yes])
+
+AS_IF([test "x$enable_nt_buffer_transfer" != xno],
+	  [AS_MESSAGE([enabling non temporal buffer transfer])
+	   AC_DEFINE([ENABLE_NT_BUFFER_TRANSFER], [1], [Enable non temporal buffer transfer])],
+	  [AC_DEFINE([ENABLE_NT_BUFFER_TRANSFER], [0], [Enable non temporal buffer transfer])]
+  )
```
i think we should try removing the compile time condition and make it runtime choice only

arun-chandran-edarath replied:
This is done to mimic "builtin-memcpy" implementation.  We can remove this entirely, but I need "#if __AVX__" around the nt-buffer-transfer code and #if __LZCNT__ around "_lzcnt_u32()"

yosefe replied:
let's keep checking AVX and LZCNT, but remove enable_nt_buffer_transfer. I think there is no reason to disable it in compile time (if have AVX and LZCNT)

arun-chandran-edarath replied:
Yes. It can be always compiled in.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -737,27 +737,25 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
-    __m256i y0, y1, y2, y3;
     size_t offset;
+    __m256i y0, y1, y2, y3;
+    const size_t switch_to_nt_store_size = 1464;
```
move if before other local vars since it's initialized

arun-chandran-edarath replied:
Done.

yosefe commented:
@tvegas1 @rakhmets can you pls review as well?

yosefe commented:
/azp run perf

tvegas1 commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -588,6 +588,28 @@ static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
+static size_t ucs_cpu_nt_dest_thresh()
+{
+    if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 5) * 29);
```
why src thresh different from dst thresh?

arun-chandran-edarath replied:
There is no "src thresh", only "dst_thresh". The variable "nt_buffer_transfer_min" is used to decide whether to use nt-buffer-transfer for a particular MPI job, not to control the selection of "ucs_x86_nt_src_buffer_transfer". It is set to "3/4" of the L3 cache size by default **mainly** to prevent the use of "nt-buffer-transfer" for Copy-in-Copy-out methods (eager-buffered) for workloads with ranks in the same L3. Users can override this by using "-x UCX_NT_BUFFER_TRANSFER_MIN=0" in the command line if their workload doesn't have ranks in the same L3.

The "dst_thresh" comes into play when the copy function "ucs_memcpy_relaxed" is called with a total_len close to the size of the L3 cache. In such cases, the hint "UCS_ARCH_MEMCPY_NT_DEST" is added based on the "dst_thresh" to avoid loading the destination data to the caches (L1, L2 and L3) and prevent cache pollution. This is because the total_len is close to the L3 cache size. The "dst_thresh" is currently set close to 90% of the L3 cache size.

Addition of hint happens here:
----------------------------------
/* Give precedence to non temporal destination */  
if (ucs_unlikely(total_len > ucs_global_opts.arch.nt_dest_threshold)) {  
    hint |= UCS_ARCH_MEMCPY_NT_DEST;  
}  

Setting of "dest_thresh":
--------------------------
static size_t ucs_cpu_nt_dest_thresh()
{
if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 5) * 29);
} else {
return UCS_MEMUNITS_INF;
}
}

When I share the osu_latency data for 4 different cases (ranks in same l3, ranks in different l3, ranks in different numa, ranks in different socket), I hope the selection criteria will become more clear.

tvegas1 replied:
thanks, so dst thresh is used to dynamically hint when nt_buffer_transfer_min has been reached.

tvegas1 commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -707,6 +732,317 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
 
+#if __AVX__
+static UCS_F_ALWAYS_INLINE
+size_t ucs_x86_nt_dst_buffer_transfer(void *dst, const void *src, size_t len,
```
given the functions, it could make sense to add simple copying tests (iterate over src aligment x iterate over length x iterate over dst alignment, for short mid and big lengts)

arun-chandran-edarath replied:
@tvegas1 Can you please elaborate? I did not understand this, please provide some examples as well? (may be tests done for builtin-memcpy?)

tvegas1 replied:
since `ucs_x86_nt_buffer_transfer()`/`ucs_x86_nt_dst_buffer_transfer()`/`ucs_x86_nt_src_buffer_transfer()` are memcpy functions now implemented in ucs, i was wondering if we could test that they are properly copying memory, possibly under `test/gtest`, with something like:
```C
<src/dest big enough 256B aligned>
for (len = 0; len < 8*1024; len++) {
  for (i = 0; i < 256; i++) {
    for (j = 0; j < 256; j++) {
      <memset pattern in src>
      ucs_x86_nt_dst_buffer_transfer(dest + i, src + i, len);
      <memcmp pattern in dst>
    }
  }
}

```

arun-chandran-edarath replied:
Done. Thanks for the suggestion.

arun-chandran-edarath replied:
@tvegas1 most of the builds are failing now with the below error due to this gtest addition. Please see the logs below.


/__w/1/s/contrib/../test/gtest/ucs/arch/test_x86_64.cc:98: undefined reference to `ucs_x86_nt_buffer_transfer'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:1429: gtest] Error 1
make[3]: Leaving directory '/home/swx-azure-svc_azpcontainer/80290-20240430.3/build/test/gtest'
make[2]: *** [Makefile:3425: all-recursive] Error 1
make[2]: Leaving directory '/home/swx-azure-svc_azpcontainer/80290-20240430.3/build/test/gtest'
make[1]: *** [Makefile:790: all-recursive] Error 1
make[1]: Leaving directory '/home/swx-azure-svc_azpcontainer/80290-20240430.3/build'
make: *** [Makefile:654: all] Error 2

Could you please help to identify the problem?

tvegas1 replied:
i think the build might be missing define `__AVX__`

arun-chandran-edarath replied:
@tvegas1 Thanks for catching it, I was always building with -mavx flag locally, so this issue never came there.

tvegas1 commented on src/uct/sm/mm/base/mm_ep.c:
```diff
@@ -320,7 +321,8 @@ static UCS_F_ALWAYS_INLINE ssize_t uct_mm_ep_am_common_send(
     case UCT_MM_SEND_AM_SHORT:
         /* write to the remote FIFO */
-        uct_am_short_fill_data(elem + 1, header, payload, length);
+        uct_am_short_fill_data(elem + 1, header, payload, length,
```
do we really have a case where length > 3/4 of Milan L3 cache (96MB CCD => 96*3/4)? else we might keep `uct_am_short_fill_data()` untouched.

arun-chandran-edarath replied:
Yes, some workloads benefit with this as explained in the above comments.  User can use "-x UCX_NT_BUFFER_TRANSFER_MIN=0" in the commandline and use nt-buffer-transfer for the entire range of transfers.

tvegas1 commented on src/ucs/arch/x86_64/global_opts.c:
```diff
@@ -17,28 +18,39 @@ ucs_config_field_t ucs_arch_global_opts_table[] = {
+   UCS_CONFIG_TYPE_MEMUNITS},
 #endif
+  {"NT_BUFFER_TRANSFER_MIN", "auto",
+   "Minimal threshold of buffer length for using nt-buffer-transfer.",
```
maybe in help replace `nt-buffer-transfer` by `non-temporal buffer`

arun-chandran-edarath replied:
Will do.

tvegas1 reviewed (commented):
just for my understanding, do we have a repro (ucx_perftest, some osu tests..) with perf gain on applicable cpu?

tvegas1 commented:
> just for my understanding, do we have a repro (ucx_perftest, some osu tests..) with perf gain on applicable cpu?

saw description now, any perftest reproducer (maybe with xpmem/shmem) on different and then same CCD?

yosefe commented:
/azp run perf

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.h:
```diff
@@ -116,7 +116,7 @@ static inline void *ucs_memcpy_relaxed(void *dst, const void *src, size_t len,
 #endif
 
-#if ((ENABLE_NT_BUFFER_TRANSFER) && (__AVX__))
+#if __AVX__
```
will change.

arun-chandran-edarath commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -736,14 +732,14 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
 }
 
-#if ((ENABLE_NT_BUFFER_TRANSFER) && (__AVX__))
+#if __AVX__
```
Will change

arun-chandran-edarath commented:
> > just for my understanding, do we have a repro (ucx_perftest, some osu tests..) with perf gain on applicable cpu?
> 
> saw description now, any perftest reproducer (maybe with xpmem/shmem) on different and then same CCD?

Please find the excel with osu_latency numbers for 4 different --map-by options (core, l3cache, numa and socket) on a zen3 machine with 32MB L3cache. CICO range is highlighted in yellow and SCOPY(xpmem) range is highlighted in green.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -595,7 +595,7 @@ static size_t ucs_cpu_nt_bt_thresh_min(size_t user_val)
 
     if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
-        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 2) * 3);
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) * 3) >> 2);
```
/4 instead of >>2

arun-chandran-edarath replied:
Done.

yosefe commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -604,7 +604,7 @@ static size_t ucs_cpu_nt_bt_thresh_min(size_t user_val)
 {
     if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
-        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) >> 5) * 29);
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) * 29) >> 5);
```
/32 instead of >>5

arun-chandran-edarath replied:
Done.

yosefe commented on test/gtest/ucs/arch/test_x86_64.cc:
```diff
@@ -63,6 +63,58 @@ class test_arch : public ucs::test {
     }
+
+#ifdef __AVX__
+    void nt_buffer_transfer_test(ucs_arch_memcpy_hint_t hint) {
```
can we make it part of 
UCS_TEST_SKIP_COND_F(test_arch, memcpy

arun-chandran-edarath replied:
I was not able to make it compile with 'UCS_TEST_SKIP_COND_F'

what about replacing nt_buffer_transfer entry function with memcpy(). I have attached a patch, please check
[gtest_fix.txt](https://github.com/openucx/ucx/files/15165975/gtest_fix.txt)

yosefe replied:
ok, let's keep as is

arun-chandran-edarath replied:
Keeping the test code as currently in test/gtest/ucs/arch/test_x86_64.cc

arun-chandran-edarath commented on test/gtest/ucs/arch/test_x86_64.cc:
```diff
@@ -102,4 +155,19 @@ UCS_TEST_SKIP_COND_F(test_arch, memcpy, RUNNING_ON_VALGRIND || !ucs::perf_retry_
     }
 }
 
+#ifdef __AVX__
```
This is better. Done. Thanks

arun-chandran-edarath commented on test/gtest/ucs/arch/test_x86_64.cc:
```diff
@@ -62,6 +63,58 @@ class test_arch : public ucs::test {
+         */
+        total_size = test_window_size + (2 * hole_size);
+
+        posix_memalign((void **)&test_window_src, 256, total_size);
```
Done.

arun-chandran-edarath commented on test/gtest/ucs/arch/test_x86_64.cc:
```diff
@@ -62,6 +63,58 @@ class test_arch : public ucs::test {
+                    /* Perform the transfer */
+                    ucs_x86_nt_buffer_transfer(dst + i, src + j, len, hint, len);
+                    result = memcmp(src + j, dst + i, len);
+                    EXPECT_EQ(result, 0);
```
Done

tvegas1 commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -588,6 +588,28 @@ static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
+        return user_val;
+    }
+
+    if ((ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD)) {
```
remove extra parenthesis here and below?

arun-chandran-edarath replied:
Done.

tvegas1 commented on src/ucs/arch/x86_64/global_opts.c:
```diff
@@ -17,28 +18,39 @@ ucs_config_field_t ucs_arch_global_opts_table[] = {
 
+    ucs_config_sprintf_memunits(min_thresh_str, sizeof(min_thresh_str),
+                                &config->nt_buffer_transfer_min, NULL);
+    printf("# Using nt-buffer-transfer for sizes from %s\n",
```
do we need to also add  nt buffer dst threshold to the printed list? else it's not possible to know what size it is

arun-chandran-edarath replied:
How to check these prints in a job started with mpirun? Any command-line param ?

tvegas1 replied:
`mpirun -np 2 ucx_info -M` should do

arun-chandran-edarath replied:
Done. Thanks for the command, it is printing as expected.

tvegas1 commented on src/ucs/sys/iovec.c:
```diff
@@ -40,10 +42,13 @@ size_t ucs_iov_copy(const struct iovec *iov, size_t iov_cnt,
+
         if (dir == UCS_IOV_COPY_FROM_BUF) {
-            memcpy(iov_buf, UCS_PTR_BYTE_OFFSET(buf, copied), len);
+            ucs_memcpy_relaxed(iov_buf, UCS_PTR_BYTE_OFFSET(buf, copied), len,
```
here we might need to hint at least with the original/non-updated length `iov[i].iov_len`?

arun-chandran-edarath replied:
@tvegas1 total_len parameter to ucs_memcpy_relaxed() should be  iov[i].iov_len ?

tvegas1 replied:
re-reading, actually using `len` here was right

tvegas1 commented on src/ucs/arch/x86_64/cpu.c:
```diff
@@ -588,6 +588,28 @@ static size_t ucs_cpu_memcpy_thresh(size_t user_val, size_t auto_val)
+    }
+
+    if (ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_AMD) {
+        return ((ucs_cpu_get_cache_size(UCS_CPU_CACHE_L3) * 3) / 4);
```
i guess we can remove all parenthesis line 598 and 607

tvegas1 reviewed (commented):
last comment

yosefe commented:
@arun-chandran-edarath code is ok but there are 2 issues with the test:
1. By default it doesn't run in CI because need to pass "--enable-optimizations" to build with AVX support. Can we add CI step for this - build with optimizations and run test_arch tests?
2. The test takes very long time to run because it iterates over (window_size,source_align,dst_align) - need to make it no longer than 10-20 seconds

arun-chandran-edarath commented:
> 
> 1. By default it doesn't run in CI because need to pass "--enable-optimizations" to build with AVX support. Can we add CI step for this - build with optimizations and run test_arch tests?

@yosefe @tvegas1 
How shall we do it?

> 2. The test takes very long time to run [[1] because it iterates over (window_size,source_align,dst_align) - need to make it no longer than 10-20 seconds
> 
I reduced the test_window_src to 4k and alignment to 64. This is enough to test all the scenarios and branches in the code.

With this I currently get the below execution times.
[ RUN      ] test_arch.nt_buffer_transfer_nt_src <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src (3036 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_dst (9485 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src_dst (11294 ms)
[----------] 4 tests from test_arch (23815 ms total)

yosefe commented:
> @yosefe @tvegas1
> How shall we do it?

See https://github.com/openucx/ucx/blob/d510358597f64bdd7c9f41a21e0dac34839d19ef/contrib/test_jenkins.sh#L1003 for example to to build with custom flags and run specific unit tests. I guess we'll need it only when running on AMD CPU?

yosefe commented:
> I reduced the test_window_size to 4k and alignment to 64. This is enough to test all the scenarios and branches in the code.

BTW, how long does the test run under ASAN?

arun-chandran-edarath commented:
> > I reduced the test_window_size to 4k and alignment to 64. This is enough to test all the scenarios and branches in the code.
> 
> BTW, how long does the test run under ASAN?

Under ASAN, it was taking longer time to execute, so I reduced test_window_size to 3K. With that, the execution times are shown below.

a) Without ASAN.
[       OK ] test_arch.memcpy (0 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src (1838 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_dst (5097 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src_dst (6077 ms)
[----------] 4 tests from test_arch (13012 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (13013 ms total)
[  PASSED  ] 4 tests.


b) With ASAN
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from test_arch, where TypeParam = 
[ RUN      ] test_arch.memcpy <> <>
[     SKIP ] (RUNNING_ON_VALGRIND || !ucs::perf_retry_count)
[       OK ] test_arch.memcpy (1 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src (4058 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_dst (5818 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src_dst (5718 ms)
[----------] 4 tests from test_arch (15595 ms total)

yosefe commented:
> > > I reduced the test_window_size to 4k and alignment to 64. This is enough to test all the scenarios and branches in the code.
> > 
> > 
> > BTW, how long does the test run under ASAN?
> 
> Under ASAN, it was taking longer time to execute, so I reduced test_window_size to 3K. With that, the execution times are shown below.
> 
> a) Without ASAN. [ OK ] test_arch.memcpy (0 ms) [ RUN ] test_arch.nt_buffer_transfer_nt_src <> <> [ OK ] test_arch.nt_buffer_transfer_nt_src (1838 ms) [ RUN ] test_arch.nt_buffer_transfer_nt_dst <> <> [ OK ] test_arch.nt_buffer_transfer_nt_dst (5097 ms) [ RUN ] test_arch.nt_buffer_transfer_nt_src_dst <> <> [ OK ] test_arch.nt_buffer_transfer_nt_src_dst (6077 ms) [----------] 4 tests from test_arch (13012 ms total)
> 
> [----------] Global test environment tear-down [==========] 4 tests from 1 test suite ran. (13013 ms total) [ PASSED ] 4 tests.
> 
> b) With ASAN [==========] Running 4 tests from 1 test suite. [----------] Global test environment set-up. [----------] 4 tests from test_arch, where TypeParam = [ RUN ] test_arch.memcpy <> <> [ SKIP ] (RUNNING_ON_VALGRIND || !ucs::perf_retry_count) [ OK ] test_arch.memcpy (1 ms) [ RUN ] test_arch.nt_buffer_transfer_nt_src <> <> [ OK ] test_arch.nt_buffer_transfer_nt_src (4058 ms) [ RUN ] test_arch.nt_buffer_transfer_nt_dst <> <> [ OK ] test_arch.nt_buffer_transfer_nt_dst (5818 ms) [ RUN ] test_arch.nt_buffer_transfer_nt_src_dst <> <> [ OK ] test_arch.nt_buffer_transfer_nt_src_dst (5718 ms) [----------] 4 tests from test_arch (15595 ms total)

Can we reduce the time 2x more, or skip the test altogether on ASAN?

arun-chandran-edarath commented:
> Can we reduce the time 2x more, or skip the test altogether on ASAN?

I reduced test_window_size to 2K. This is the least minimum size required, it covers the case 'switch_to_nt_store_size' which is currently at '1464'. We cannot go much further.

With that + ASAN,  I get the execution time as shown below (total time of tests is less than 10s).

[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from test_arch, where TypeParam =
[ RUN      ] test_arch.memcpy <> <>
[     SKIP ] (RUNNING_ON_VALGRIND || !ucs::perf_retry_count)
[       OK ] test_arch.memcpy (0 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src (2131 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_dst (2678 ms)
[ RUN      ] test_arch.nt_buffer_transfer_nt_src_dst <> <>
[       OK ] test_arch.nt_buffer_transfer_nt_src_dst (2646 ms)
[----------] 4 tests from test_arch (7455 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (7455 ms total)
[  PASSED  ] 4 tests.

arun-chandran-edarath commented:
> 1. --enable-optimizations

@yosefe 
Could you please check the attached patch for the CI test implementation? 
[jenkin.patch.txt](https://github.com/openucx/ucx/files/15220318/jenkin.patch.txt)

yosefe commented:
> > 1. --enable-optimizations
> 
> @yosefe Could you please check the attached patch for the CI test implementation? [jenkin.patch.txt](https://github.com/openucx/ucx/files/15220318/jenkin.patch.txt)

Looks OK, but do we really need to enable ASAN?

yosefe commented on contrib/test_jenkins.sh:
```diff
@@ -1051,6 +1051,18 @@ run_release_mode_tests() {
+#
+run_nt_buffer_transfer_tests() {
+    lscpu_output=$(lscpu)
+    if [[ $lscpu_output == *'AuthenticAMD'*  ]]; then
```
minor:
```
if lscpu | grep -q 'AuthenticAMD'
then
        build release --enable-gtest --enable-optimizations
        ...
fi
```

arun-chandran-edarath replied:
Done. Thanks

arun-chandran-edarath commented:
@yosefe
Do we need to change the selection criteria of "builtin_memcpy"? Currently, it is based on 'len', more accurate is 'total_len' right?

`diff --git a/src/ucs/arch/x86_64/cpu.h b/src/ucs/arch/x86_64/cpu.h
index 5fe5ada19..77248e834 100644
--- a/src/ucs/arch/x86_64/cpu.h
+++ b/src/ucs/arch/x86_64/cpu.h
@@ -102,8 +102,8 @@ static inline void *ucs_memcpy_relaxed(void *dst, const void *src, size_t len,
                                        size_t total_len)
 {
 #if ENABLE_BUILTIN_MEMCPY
-    if (ucs_unlikely((len > ucs_global_opts.arch.builtin_memcpy_min) &&
-                     (len < ucs_global_opts.arch.builtin_memcpy_max))) {
+    if (ucs_unlikely((toral_len > ucs_global_opts.arch.builtin_memcpy_min) &&
+                     (toral_len < ucs_global_opts.arch.builtin_memcpy_max))) {
         asm volatile ("rep movsb"
                       : "=D" (dst),
                       "=S" (src),
`

tvegas1 commented:
> @yosefe Do we need to change the selection criteria of "builtin_memcpy"? Currently, it is based on 'len', more accurate is 'total_len' right?
> 
> `diff --git a/src/ucs/arch/x86_64/cpu.h b/src/ucs/arch/x86_64/cpu.h index 5fe5ada19..77248e834 100644 --- a/src/ucs/arch/x86_64/cpu.h +++ b/src/ucs/arch/x86_64/cpu.h @@ -102,8 +102,8 @@ static inline void *ucs_memcpy_relaxed(void *dst, const void *src, size_t len, size_t total_len) { #if ENABLE_BUILTIN_MEMCPY
> 
> * if (ucs_unlikely((len > ucs_global_opts.arch.builtin_memcpy_min) &&
> * ```
>                   (len < ucs_global_opts.arch.builtin_memcpy_max))) {
>   ```
> 
> * if (ucs_unlikely((toral_len > ucs_global_opts.arch.builtin_memcpy_min) &&
> * ```
>                   (toral_len < ucs_global_opts.arch.builtin_memcpy_max))) {
>      asm volatile ("rep movsb"
>                    : "=D" (dst),
>                    "=S" (src),
>   ```
> 
> `

imo using `len` here is fine, since it is done when not in NT-buffer context

arun-chandran-edarath commented:
> > @yosefe Do we need to change the selection criteria of "builtin_memcpy"? Currently, it is based on 'len', more accurate is 'total_len' right?
> > `diff --git a/src/ucs/arch/x86_64/cpu.h b/src/ucs/arch/x86_64/cpu.h index 5fe5ada19..77248e834 100644 --- a/src/ucs/arch/x86_64/cpu.h +++ b/src/ucs/arch/x86_64/cpu.h @@ -102,8 +102,8 @@ static inline void *ucs_memcpy_relaxed(void *dst, const void *src, size_t len, size_t total_len) { #if ENABLE_BUILTIN_MEMCPY
> > 
> > * if (ucs_unlikely((len > ucs_global_opts.arch.builtin_memcpy_min) &&
> > * ```
> >                   (len < ucs_global_opts.arch.builtin_memcpy_max))) {
> >   ```
> > * if (ucs_unlikely((toral_len > ucs_global_opts.arch.builtin_memcpy_min) &&
> > * ```
> >                   (toral_len < ucs_global_opts.arch.builtin_memcpy_max))) {
> >      asm volatile ("rep movsb"
> >                    : "=D" (dst),
> >                    "=S" (src),
> >   ```
> > 
> > `
> 
> imo using `len` here is fine, since it is done when not in NT-buffer context

Usually, vendors recommend to use ERMS memcpy between l1 and l2.  But the 'len' in ucs_memcpy_relaxed() is not the actual length of transfer, it is total_len/segment_size. 
Eg:
    [UCS_CPU_VENDOR_INTEL] = {
        .min = 1 * UCS_KBYTE,
        .max = 8 * UCS_MBYTE
    },
using 'len' will end up in using builtin_memcpy() for a total_len of 9MB also right? because that transfer is usually divided into chunks of 256KB.

yosefe commented:
Let's not change builtin_memcpy logic un this PR.

tvegas1 commented:
> using 'len' will end up in using builtin_memcpy() for a total_len of 9MB also right? because that transfer is usually divided into chunks of 256KB.

i see, but maybe between two segments copying, cache has been replaced enough that it's more accurate to take `len`. also this would be separate change i guess.

arun-chandran-edarath commented:
@yosefe, it appears that there are currently 29 failing tests. I'm unsure if my CI test commit is the cause of these failures. Could you please check?

yosefe commented:
@arun-chandran-edarath the failures indeed seem unrelated. Can you pls squash the commits to a single commit for now (preferably, without rebasing) and we will check the CI issues.

arun-chandran-edarath commented:
> @arun-chandran-edarath the failures indeed seem unrelated. Can you pls squash the commits to a single commit for now (preferably, without rebasing) and we will check the CI issues.
I squashed all the commits into a single one. Please check. Thanks.
